lib/timers.js

```diff
@@ -440,13 +452,20 @@ function rearm(timer) {
 
 
 const clearTimeout = exports.clearTimeout = function(timer) {
-  if (timer && (timer[kOnTimeout] || timer._onTimeout)) {
+  if (!timer) return;
+
+  if (timer[kOnTimeout] || timer._onTimeout) {
+    if (typeof timer._onTimeout !== 'function') {
+      promiseResolve(timer._onTimeout, undefined);
+    }
     timer[kOnTimeout] = timer._onTimeout = null;
     if (timer instanceof Timeout) {
       timer.close(); // for after === 0
     } else {
       unenroll(timer);
     }
+  } else if (isPromise(timer)) {
```

Comment:
This seems like a bad idea for many reasons. I think it's fine if you can't cancel promisified timeout until we have some generic cancellation mechanism.